src/main/java/hudson/matrix/MatrixBuild.java

```diff
@@ -91,34 +91,32 @@ public Object readResolve() {
     }
 
     /**
+     * Delete build and its child builds.
+     *
+     * Child builds can be skipped if they have to be kept.
+     *
+     * @since TODO
+     */
+    @Override
+    public void delete() throws IOException {
+        super.delete();
+
+        for(MatrixRun run : getExactRuns()) {
+            if (run.getWhyKeepLog() != null) continue;
```

Comment:
`MatrixRun`.`getWhyKeepLog` can be non-null while  `MatrixBuild`.`getWhyKeepLog` is null in case it is locked as the last completed build of its configuration. My intention here is to remove parent and all the children that are not locked. (I am not a fan of this `MatrixRun` locking so I am opened to delete them all).